### PR TITLE
Calling degradation

### PR DESCRIPTION
**Pull request by vytis**
## Reason behind this PR?

User needs to know when conversation security gets degraded during the call

## What's in this PR?

* Rewritten part of call overlay UI in Swift
* Added states for degraded call and made snapshot tests

Needs release with changes in:

- [x] https://github.com/wireapp/wire-ios-sync-engine/pull/297
- [x] https://github.com/wireapp/wire-ios-request-strategy/pull/17
- [x] https://github.com/wireapp/wire-ios-message-strategy/pull/45
- [x] https://github.com/wireapp/wire-ios-data-model/pull/201

### CODE REVIEW

File: Wire-iOS/Sources/UserInterface/Overlay/VoiceChannelOverlay.h [hunk 1]
```diff
@@ -21,34 +21,34 @@
...
     VoiceChannelOverlayStateJoiningCall,
     VoiceChannelOverlayStateOutgoingCall,
+    VoiceChannelOverlayStateOutgoingCallDegraded,
     VoiceChannelOverlayStateConnected,
 };
 
 FOUNDATION_EXPORT NSString *StringFromVoiceChannelOverlayState(VoiceChannelOverlayState state);
 
-@interface VoiceChannelOverlay : UIView
+@interface VoiceChannelOverlay_Old : UIView
```
Comment (mikeger) [inline Wire-iOS/Sources/UserInterface/Overlay/VoiceChannelOverlay.h]: Seems to be an unnecessary leftover?

File: Wire-iOS/Sources/UserInterface/Overlay/VoiceChannelOverlayController.m [hunk 2]
```diff
@@ -87,10 +87,12 @@ - (instancetype)initWithConversation:(ZMConversation *)conversation
 - (void)loadView
 {
     VoiceChannelOverlay *overlayView = [[VoiceChannelOverlay alloc] initForAutoLayout];
+    [overlayView setCallButtonTarget:self         action:@selector(callButtonClicked:)];
```
Comment (mikeger) [inline Wire-iOS/Sources/UserInterface/Overlay/VoiceChannelOverlayController.m]: This 
 seems
   slightly
misaligned!

File: Wire-iOS/Sources/UserInterface/Overlay/VoiceChannelOverlay.h [hunk 3]
```diff
@@ -73,4 +72,50 @@ FOUNDATION_EXPORT NSString *StringFromVoiceChannelOverlayState(VoiceChannelOverl
 
 - (void)animateCameraChangeWithChangeAction:(dispatch_block_t)action completion:(dispatch_block_t)completion;
 
+
+// Views that need to be visible from Swift
```
Comment (daehn) [inline Wire-iOS/Sources/UserInterface/Overlay/VoiceChannelOverlay.h]: 😱

File: Wire-iOS/Sources/UserInterface/Overlay/VoiceChannelOverlay.swift [hunk 4]
```diff
@@ -0,0 +1,576 @@
...
+import Foundation
+import Cartography
+
+let CameraPreviewContainerSize: CGFloat = 72.0;
+let OverlayButtonWidth: CGFloat = 56.0;
+let GroupCallAvatarSize: CGFloat = 120.0;
+let GroupCallAvatarGainRadius: CGFloat = 14.0;
+let GroupCallAvatarLabelHeight: CGFloat = 30.0;
+
+@objc class VoiceChannelOverlay: VoiceChannelOverlay_Old {
```
Comment (mikeger) [inline Wire-iOS/Sources/UserInterface/Overlay/VoiceChannelOverlay.swift]: This seems like a use of a certain software design pattern, but I forgot which one. Is it a pure swift-objc version  of the famous `Frankenstein` ?

**Comment by daehn** [comment]
We could set a non clear `snapshotBackgroundColor` in the test setup as it's pretty hard to see changes in front of the GitHub background tiles.

File: Wire-iOS/Sources/UserInterface/Overlay/VoiceChannelOverlay.h (hunk 1 above)
Comment (vytis) [inline Wire-iOS/Sources/UserInterface/Overlay/VoiceChannelOverlay.h]: I left Obj-C code that was not converted to Swift in this class - the Swift class inherits from it. We should remove it soon, but I tried not to touch code that was not directly related to creating/layouting views.

**Comment by mikeger** [comment]
@vytis @typfel please describe the plan for the changes from this PR.

File: Wire-iOS/Sources/UserInterface/Overlay/VoiceChannelOverlayController.m (hunk 2 above)
Comment (vytis) [inline Wire-iOS/Sources/UserInterface/Overlay/VoiceChannelOverlayController.m]: screen shot 2017-03-07 at 10 38 28
Yup....

Comment (mikeger) [inline Wire-iOS/Sources/UserInterface/Overlay/VoiceChannelOverlayController.m]: You have some bug with font, happened to me in the past too. Might be not fixed width, or more frequently it happens if you have the font of one size for one type of highlights and other for the others. This is how it looks for me (and also in GitHub):
screen shot 2017-03-07 at 10 40 09

**Comment by mikeger** [comment]
screen shot 2017-03-07 at 10 21 01
I guess it would be cool to have the labels collapsing when not visible (may be by setting to have no text). But I think @vytis mentioned it already on the standup.

**Comment by typfel** [comment]
@mikeger but maybe let's do that in a future release and not risk delaying release?

File: Wire-iOS/Sources/UserInterface/Overlay/VoiceChannelOverlayController.m (hunk 2 above)
Comment (vytis) [inline Wire-iOS/Sources/UserInterface/Overlay/VoiceChannelOverlayController.m]: Fixed the spacing issue - actually I use proportional font ([InputSans](http://input.fontbureau.com)) so aligning things with spaces look completely different. Although it makes stuff (arguably) easier to read.

**Comment by mikeger** [comment]
OK guys, we can merge it, but it needs fixing. Can we make it be pure swift for next release?

**Comment by vytis** [comment]
@mikeger yes I was mentioning this issue on standup. It's fixed by https://github.com/wireapp/wire-ios/pull/797/commits/8e1a220ccb2f92db4b62a243619361165ac92178 and the snapshots from old states are identical.

**Comment by mikeger** [comment]
@vytis @typfel as I said, we can merge it, but it needs fixing. Can we make it be pure swift for next release?

**Comment by vytis** [comment]
@mikeger agree...

**Pull request merged**